### ISSUE TITLE
Variance of the gradients is with respect to scaled down gradients

### ISSUE THREAD

**Issue by ptigas** [f-dangel/backpack#49]
Hi,

I was trying to estimate the variance of the gradients and I observed the following. It seems that the variance is not with respect to the actual gradients but the scaled-down by batch-size version of them. Here's a quick example to illustrate this:

```
B = 20 # Batch size

# Create a simple NN
m = nn.Sequential(nn.Linear(10, 32),nn.ReLU(), nn.Linear(32, 1),)

# And a dummy loss - this needs be a nn.Module for backpack to work
class Loss(nn.Module):
  def forward(self, x):
    return x.mean()

loss = extend(Loss())
m = extend(m)
batch = torch.rand(B, 10)
```

Groundtruth variance is estimated by taking the gradients per example and computing the gradients
```
clear_backprops(m)
gradients = []
for i in range(B):
  m.zero_grad()
  loss(m(batch)[i]).backward(retain_graph=True)
  gradients.append(torch.cat([g.grad.view(-1) for g in list(m.parameters())], dim=0))
gradients = torch.stack(gradients)
ground_truth_variance = gradients.var(0)
print(gradients.var(0).mean())
```

Here's what backpack returns
```
m.zero_grad()
with backpack(extensions.Variance()):
    loss(m(batch)).backward()

grad_vars = torch.cat([g.variance.view(-1) for g in list(m.parameters())])

print(grad_vars.mean())
assert np.allclose(ground_truth_variance, grad_vars)
```

And here's backpack after I scale them back:
```
m.zero_grad()
with backpack(extensions.BatchGrad()):
    loss(m(batch)).backward()
grad_vars = torch.cat([(g.grad_batch*B).var(0).view(-1) for g in list(m.parameters())])

print(grad_vars.mean())
assert np.allclose(ground_truth_variance, grad_vars)
```

Let me know what you think
p.

**Comment by fKunstner** [comment]
Hi @ptigas, thanks for reaching out! You're right, there's something fishy here. We need to at least improve the docs. 

Summarizing the problem, it's the difference in behavior between the two following calls;
```
with(Backpack(Variance())):
    loss(model(x), y, reduction="mean").backward()         # Using the mean
with(Backpack(Variance())):
    loss(model(x), y, reduction="sum").backward()          # Using the sum
```

Writing `fᵢ` for the loss of the `ith` sample and `gᵢ` for its gradient, with `i = 1, …, n`, without Backpack, the first call computes the gradient of  `¹/ₙ ∑ᵢ fᵢ`  and the second of `∑ᵢ fᵢ`.

The Variance functionality receives vectors, `[v₁, …, vₙ]`, and computes their element-wise variance. Using the mean or the sum changes what information gets propagated, and the output. If the loss is a sum, the variance function receives `[v₁, …, vₙ] = [g₁, …, gₙ]`. If the loss is a mean, it is `[v₁, …, vₙ] = [¹/ₙ g₁, …, ¹/ₙ gₙ]`. The variance for the second one will be a `1/n²` -scaled version of the first one.

Hopefully this makes sense. We're going to add this to the doc in a release soon. 

---

Other than updating the docs, I'm not sure yet what's the best way to address this. 

Most users probably other expect to get the variance of `[g₁, …, gₙ]`, regardless of the reduction used. I don't think we can do this automatically though. We would like the first-order extensions to work with arbitrary loss functions (without the need to extend them), but we wouldn't have a way to know what the loss does.

Maybe we could add a `multiply_by_batch` option, which would multiply by `n` before computing the variance. We could make it `True` by default to match the default reduction in PyTorch, but users would have to switch it to `False` if they're using a sum, which isn't ideal. This would work like
```
with backpack(Variance(multiply_by_batch=False)):
    loss.backward()
```

Open to suggestions

**Comment by fKunstner** [comment]
Addressed in the [documentation](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.BatchGrad) in [1.1.1](https://github.com/f-dangel/backpack/releases/tag/1.1.1)
